src/supa/nrm/backend.py

```diff
@@ -0,0 +1,96 @@
+#  Copyright 2022 SURF.
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import structlog
+from sqlalchemy.orm import aliased, session
+from structlog.stdlib import BoundLogger
+
+from supa.db.model import Port, Reservation
+
+logger = structlog.get_logger(__name__)
+
+
+class BaseBackend:
+    """Default backend interface to Network Resource Manager.
+
+    Backend interface between the following NSI primitives and the local NRM:
+
+    reserve, reserve_timeout, reserve_commit, reserve_abort, provision, release, activate, deactivate and terminate
+
+    The arguments for all functions are the same, for example the reserve():
+
+    def reserve(
+        self, connection_id: UUID, src_port_id: str, src_vlan: int, dst_port_id: str, dst_vlan: int, bandwidth: int
+    ) -> None:
+        ...
+
+    If a function for a primitive is not defined the call the NRM will be skipped.
+    """
+
+    log: BoundLogger
+
+    def __init__(self) -> None:
+        """Initialize the BaseBackend."""
+        self.log = logger.bind(backend="no-op")
+
+
+"""Set backend to BaseBackend which effectively does nothing,
+but this can be overwritten by a custom backend
+as declared in supa.env or on the command line of `supa serve`"""
+backend = BaseBackend()
+
+
+def call_backend(primitive: str, reservation: Reservation, database_session: session) -> None:
```

Comment:
Perhaps make the call_backend function a Not implemented BaseBackend function? And then implement it in the actual class.